temporalio/nexus/handler/_operation_context.py

```diff
@@ -0,0 +1,351 @@
+from __future__ import annotations
+
+import contextvars
+import logging
+import re
+import urllib.parse
+from contextvars import ContextVar
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Mapping,
+    Optional,
+    Sequence,
+    Union,
+)
+
+import nexusrpc.handler
+from nexusrpc.handler import CancelOperationContext, StartOperationContext
+
+import temporalio.api.common.v1
+import temporalio.api.enums.v1
+import temporalio.common
+from temporalio.nexus.handler._token import WorkflowOperationToken
+from temporalio.types import (
+    MethodAsyncSingleParam,
+    ParamType,
+    ReturnType,
+    SelfType,
+)
+
+if TYPE_CHECKING:
+    from temporalio.client import Client, WorkflowHandle
+
+
+logger = logging.getLogger(__name__)
+
+
+_current_context: ContextVar[TemporalOperationContext] = ContextVar(
+    "temporal-operation-context"
+)
+
+
+@dataclass
+class TemporalOperationContext:
+    """
+    Context for a Nexus operation being handled by a Temporal Nexus Worker.
+    """
+
+    nexus_operation_context: Union[StartOperationContext, CancelOperationContext]
+
+    client: Client
+    """The Temporal client in use by the worker handling this Nexus operation."""
+
+    task_queue: str
+    """The task queue of the worker handling this Nexus operation."""
+
+    @staticmethod
+    def try_current() -> Optional[TemporalOperationContext]:
+        return _current_context.get(None)
+
+    @staticmethod
+    def current() -> TemporalOperationContext:
+        context = TemporalOperationContext.try_current()
+        if not context:
+            raise RuntimeError("Not in Nexus operation context")
+        return context
+
+    @staticmethod
+    def set(
+        context: TemporalOperationContext,
+    ) -> contextvars.Token[TemporalOperationContext]:
+        return _current_context.set(context)
+
+    @staticmethod
+    def reset(token: contextvars.Token[TemporalOperationContext]) -> None:
+        _current_context.reset(token)
+
+    @property
+    def temporal_start_operation_context(
+        self,
+    ) -> Optional[_TemporalStartOperationContext]:
+        ctx = self.nexus_operation_context
+        if not isinstance(ctx, StartOperationContext):
+            return None
+        return _TemporalStartOperationContext(ctx)
+
+    @property
+    def temporal_cancel_operation_context(
+        self,
+    ) -> Optional[_TemporalCancelOperationContext]:
+        ctx = self.nexus_operation_context
+        if not isinstance(ctx, CancelOperationContext):
+            return None
+        return _TemporalCancelOperationContext(ctx)
+
+    # Overload for single-param workflow
+    # TODO(nexus-prerelease): support other overloads?
+    async def start_workflow(
+        self,
```

Comment:
This is the special Nexus `start_workflow` that starts the "backing workflow", setting the callback, sets bidi-links, propagates request ID etc.

Notice that it does not return `WorkflowHandle`: it returns `WorkflowOperationToken`. This prevents users from accidentally using `client.start_workflow`: there will be a type error, and there will be a request time error, although this will be after the workflow has started. We can't make it error before the request is issued, because we want to allow starting normal workflows via `client.start_workflow`.

It can only be called with an active Start Operation context, so I think it makes sense for it to be accessed as a method on the context.